Update round(::Missing) to match float version

amilsted opened the pull request:
`round()` was updated to accept `digits` and `base` as keyword arguments. It looks like the `Missing` version was... missed? 😈

This prevents e.g. `round(A, digits=4)` and `round(A, 4)` (old signature) from working on `Union{Missing, Float64}` arrays `A`.

Fixes #31076

amilsted committed:
Update round(::Missing) to match float version

`round()` was updated to accept `digits` and `base` as keyword arguments. It looks like the `Missing` version was... missed? 😈
This prevents e.g. `round(A, digits=4)` and `round(A, 4)` (old signature) from working on `Union{Missing, Float64}` arrays `A`.

amilsted committed:
Fix tests for updated round(::Missing, ...)

amilsted committed:
Test elementwise rounding on Union{T, Missing}[...]

When the signature of `Base.round()` and friends (ceil,floor,trunc) was changed to accept keyword arguments for `digits` and `base` instead of positional arguments, the implementations for `Missing` were missed. Test these functions elementwise on arrays of Floats and Missing values to make sure their signatures are compatible with the versions for numbers.

musm commented on base/missing.jl:
```diff
@@ -109,7 +109,7 @@ max(::Any,     ::Missing) = missing
 for f in (:(ceil), :(floor), :(round), :(trunc))
     @eval begin
-        ($f)(::Missing, digits::Integer=0, base::Integer=0) = missing
+        ($f)(::Missing; digits::Integer=0, base=0) = missing
```
why remove `base::Integer` ?

amilsted replied:
Just trying to sync with the docstring [here](https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/floatfuncs.jl#L47-L120). I guess non-integer bases are, in principle, a [thing](https://en.wikipedia.org/wiki/Non-integer_representation), and the `Missing` implementation will always return `missing` anyway. Happy to put `::Integer` back in if that's preferred.

musm replied:
AFAIK, the respective julia functions don't support non-integer bases, so it would be best to leave it so that a proper method error is thrown.

nalimilan replied:
I suspect the type was just omitted from the docstring for simplicity.

nalimilan commented on test/missing.jl:
```diff
@@ -193,11 +193,15 @@ end
         @test ismissing(f(Union{Int, Missing}, missing))
         @test f(Union{Int, Missing}, 1.0) === 1
         @test_throws MissingException f(Int, missing)
+        @test ismissing.(f.([1.0, missing])) == [false, true]
```
I don't think these tests are useful. `f` is called either on `missing` which is covered above, or on `1.0` which is covered elsewhere. Please remove them to keep the tests simple and similar to what is done for other operations.

amilsted replied:
Do you just mean line 201, or lines 201-204? The reason I added these tests was to prevent the Number and Missing versions of these functions from getting out of sync again, so I call them elementwise with the various argument combinations. If the Number definitions change again, this *might* catch it. This is how I interpreted the request to "exercise this codepath", but I'm happy to drop these tests if they're unwelcome. I guess we're unlikely to see the same problem again!

nalimilan replied:
OK, let's keep these then. I'm afraid other functions in this file would need the same treatment, but I guess covering part of them is better than nothing.

amilsted replied:
For the record, not testing any given Missing functionality on mixed arrays is rather like not testing a car on roads. It looks like there are a lot of other elementwise tests in the file.

nalimilan replied:
Writing good tests is very hard. Ideally we should only test the elementary blocks to avoid combinatorial explosion of tests. I can be a good idea to test a few common patterns too, but in the present case the `f.(...)` really just applies `f` to each element, so that's nothing more than a compact way of applying `f` to each entry and checking the result.

amilsted replied:
Yes, but if you only test these functions on scalars, you can get the problem that led to this PR: The signature of `round(x::Float...)` changed, but `round(x::Missing...)` was forgotten (and so were its tests, naturally). Because tests on mixed `Union{Float..., Missing}` arrays did not exist, this problem did not show up as a test failure. Since the whole point of `Missing` is to support these mixed arrays, I think that was a gap in the tests, which these new tests fill in. Note that there are many other tests for mixed arrays in the `test/missing.jl`.

nalimilan replied:
Again, arrays are *not* involved in this code. What matters is that `round` is called with the same arguments on both `missing` and non-`missing` values. We could do that on separate lines if we wanted, the result would be the same.

amilsted replied:
Yes, I understand that, but isn't it nice to have it in one line? 😃 More seriously, it would seem weird (to me) to check the `Float` function twice, here and in the float tests (although I realize this is what's happening behind the scenes). On the other hand, it seems worthwhile to test that the signatures are compatible in order to avoid the kind of problem that triggered this PR. I thought this would be a good way to do that. Do you know of a nicer way to do this? 

Perhaps I should add a comment explaining why these tests are present...

amilsted replied:
I have removed duplicated tests of Missing functionality and added some comments.

nalimilan replied:
I'm not opposed to doing this that way, I just want to make it clear what that line does.

amilsted replied:
Okay, I see. Please check the latest commit, which I think cleans things up quite a bit.

amilsted committed:
Only accept Integer base in rounding

amilsted commented:
Just realized there are other variants of `round()` that should be addressed. Will update soon!

amilsted committed:
Cover missing Missing variants of round() etc.

Accept `sigdigits` keyword argument in addition to `digits` and `base`. Also accept `RoundingMode` positional argument.

nalimilan commented on base/missing.jl:
```diff
@@ -107,18 +107,22 @@ max(::Missing, ::Any)     = missing
+round(::Type{T}, x::Rational, ::RoundingMode=RoundNearest) where {T>:Missing} = round(nonmissingtype(T), x)
+round(::Type{T}, x::Rational{Bool}, ::RoundingMode=RoundNearest) where {T>:Missing} = round(nonmissingtype(T), x)
+
+trunc(::Missing; kwargs...) = round(missing, RoundToZero; kwargs...)
```
These should just return `missing`. And better keep the loop.

amilsted replied:
Updated: Didn't want to just return missing, since then arbitrary kwargs and types are allowed.

nalimilan commented on base/missing.jl:
```diff
@@ -107,18 +107,22 @@ max(::Missing, ::Any)     = missing
+floor(::Missing; kwargs...) = round(missing, RoundDown; kwargs...)
+ ceil(::Missing; kwargs...) = round(missing, RoundUp; kwargs...)
+
+trunc(::Type{T}, ::Missing; kwargs...) where {T} = round(T, missing, RoundToZero; kwargs...)
```
These don't correspond to existing methods.

amilsted replied:
Fixed.

amilsted committed:
More variants for round(::Missing) tests

Test that there are Missing versions of remaining variants of round() and friends.

amilsted committed:
Remove extraneous kwargs from trunc,floor,ceil

amilsted committed:
Clean up trunc,floor,ceil defs for Missing

Continue to call `round()`, despite it just returning `missing`, so that errors happen with kwargs or types that are invalid.

nalimilan commented on base/missing.jl:
```diff
@@ -107,16 +107,19 @@ max(::Missing, ::Any)     = missing
-        # to fix ambiguities
-        ($f)(::Type{T}, x::Rational) where {T>:Missing} = $f(nonmissingtype(T), x)
-        ($f)(::Type{T}, x::Rational{Bool}) where {T>:Missing} = $f(nonmissingtype(T), x)
+        $(f)(::Missing; kwargs...) = round(T, missing; kwargs...)
```
That's still completely incorrect. Please run tests locally before pushing, as the full test suite is run on 6 workers for each push, taking hours of CPU time.

amilsted replied:
Sorry! I didn't expect this to turn into so much work and have been doing this in a combination of REPL and editing on Github! Have run the code locally again - seems to work for me.

nalimilan replied:
I canceled them manually. GitHub editing should really be reserved to trivial changes. Every time I've used it for more complex things I've regretted it.

amilsted replied:
Yeah... this started off simple... guess I should have switched methods when it got complicated!

amilsted committed:
Fix stupid typo

amilsted committed:
Fix other stupid typo.

nalimilan commented on base/missing.jl:
```diff
@@ -107,16 +107,19 @@ max(::Missing, ::Any)     = missing
-        ($f)(::Type{T}, x::Rational) where {T>:Missing} = $f(nonmissingtype(T), x)
-        ($f)(::Type{T}, x::Rational{Bool}) where {T>:Missing} = $f(nonmissingtype(T), x)
+        ($f)(::Missing; kwargs...) = round(missing; kwargs...)
+        ($f)(::Type{T}, ::Missing) where {T} = round(T, missing)
```
Honestly it would be much clearer to list the supported keyword arguments here.

amilsted replied:
I was emulating how things are defined in float.jl. The advantage of this, I suppose, is that we only need to specify the allowed keywords for `round()`, which also needs to support the `RoundingMode` argument, so can't be taken care of in a loop with the other functions.

nalimilan replied:
But the difference is that in float.jl there's an additional `::RoundingMode` argument. Here it's really confusing to call `round` from other functions.

Anyway if you define a method for `round`, better put it in the loop so that it's defined for all functions. You can define additional methods for `round` outside of the loop if needed.

amilsted replied:
I have now returned to the original loop and just split out round() so the `::RoundingMode` argument can be taken care of.

nalimilan commented on base/missing.jl:
```diff
@@ -107,16 +107,19 @@ max(::Missing, ::Any)     = missing
-        ($f)(::Type{>:Missing}, ::Missing) = missing
-        ($f)(::Type{T}, ::Missing) where {T} =
-            throw(MissingException("cannot convert a missing value to type $T: use Union{$T, Missing} instead"))
-        ($f)(::Type{T}, x::Any) where {T>:Missing} = $f(nonmissingtype(T), x)
```
Where have these functions gone?

amilsted replied:
Good point.

amilsted replied:
I thought I had taken care of them with the wrapper that calls round(), but this didn't take care of all the possible cases. Tests now pass again for me.

amilsted committed:
Fix definitions of ceil, floor, and trunc

My attempt to avoid redefining all these cases by deferring to round() didn't work, so just define them separately.

nalimilan commented on base/missing.jl:
```diff
@@ -107,9 +107,19 @@ max(::Missing, ::Any)     = missing
+round(::Type{>:Missing}, ::Missing, ::RoundingMode=RoundNearest) = missing
+round(::Type{T}, ::Missing, ::RoundingMode=RoundNearest) where {T} =
+    throw(MissingException("cannot convert a missing value to type $T: use Union{$T, Missing} instead"))
+round(::Type{T}, x::Any, ::RoundingMode=RoundNearest) where {T>:Missing} = round(nonmissingtype(T), x)
```
`RoundingMode` should be passed to `round`. Same below. This should be covered by tests which fail without that change.

amilsted replied:
Oh dear. Yes.

nalimilan commented on test/missing.jl:
```diff
@@ -188,16 +188,23 @@ Base.one(::Type{Unit}) = 1
-        @test ismissing(f(missing, 1, 1))
+        @test ismissing(f(missing, sigdigits=1, digits=1, base=1))
         @test ismissing(f(Union{Int, Missing}, missing))
         @test f(Union{Int, Missing}, 1.0) === 1
```
This test should also have a variant with keyword arguments, passing it a value which allows testing that the result has the intended number of digits. That way we're sure the arguments are actually passed.

amilsted replied:
The versions which accept keyword arguments never call round(). They only need to accept them so that elementwise works on mixed arrays.

nalimilan replied:
No, that's not how it works. Element-wise operation just calls the method corresponding to a given element. It doesn't depend on the fact that they are stored in an array at all.

And something like `round(Union{Int, Missing}, 1.11111, digits=2)` *has* to pass keyword arguments to `round` down the stream.

amilsted replied:
> No, that's not how it works. Element-wise operation just calls the method corresponding to a given element. It doesn't depend on the fact that they are stored in an array at all.

I realize that. That's why `round(::Missing; sigdigits::Integer=1, ...)` doesn't need to call `round()` and just returns `missing`.

> And something like `round(Union{Int, Missing}, 1.11111, digits=2)` _has_ to pass keyword arguments to `round` down the stream.

That's not a valid call to `round()`. The type argument doesn't mix with the keyword arguments (although perhaps it should, especially in the case of `sigdigits`...).

nalimilan replied:
Ah, sorry, there are so many signatures.

amilsted replied:
Tell me about it...

amilsted committed:
Pass RoundingMode to round() in Missing versions

amilsted committed:
Check RoundingMode is being passed on for Missing

amilsted commented:
I'm guessing I can't cancel the builds myself? Only the latest commit needs testing.

nalimilan commented:
I don't think you can, unfortunately.

nalimilan committed:
Use isequal and ===

amilsted committed:
Test round() with sigdigits on missing-value arrays

Also, separate tests for digits and sigdigits: Combining these arguments is not valid for floats, although it currently goes through for Missing. Finally, use a sensible base for the digits and sigdigits tests.

amilsted commented:
Failed a test! Method ambiguities I don't know how to resolve... I kept the explicit `Rational` variants of `round()` in missing.jl from the original, but it seems they're not enough with the `::RoundingMode` argument. Anyone know what to do?
```
  Expression: detect_ambiguities(Core, Base; imported=true, recursive=true, ambiguous_bottom=false) == []
   Evaluated: Tuple{Method,Method}[(round(::Type{T}, x::Rational, r::RoundingMode) where T>:Missing in Base at missing.jl:116, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:NearestTiesAway}) where {T, Tr} in Base at rational.jl:390), (round(::Type{T}, x, r::RoundingMode) where T>:Missing in Base at missing.jl:114, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:Nearest}) where {T, Tr} in Base at rational.jl:374), (round(::Type{T}, x::Rational{Bool}, r::RoundingMode) where T>:Missing in Base at missing.jl:117, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:NearestTiesUp}) where {T, Tr} in Base at rational.jl:404), (round(::Type{T}, x::Rational{Bool}, r::RoundingMode) where T>:Missing in Base at missing.jl:117, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:Nearest}) where {T, Tr} in Base at rational.jl:374), (round(::Type{T}, x, r::RoundingMode) where T>:Missing in Base at missing.jl:114, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesAway}) where T in Base at rational.jl:425), (round(::Type{T}, x::Rational{Bool}, r::RoundingMode) where T>:Missing in Base at missing.jl:117, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:Nearest}) where T in Base at rational.jl:424), (round(::Type{T}, x, r::RoundingMode) where T>:Missing in Base at missing.jl:114, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:Nearest}) where T in Base at rational.jl:424), (round(::Type{T}, x, r::RoundingMode) where T>:Missing in Base at missing.jl:114, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:NearestTiesUp}) where {T, Tr} in Base at rational.jl:404), (round(::Type{T}, x::Rational, r::RoundingMode) where T>:Missing in Base at missing.jl:116, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:NearestTiesUp}) where {T, Tr} in Base at rational.jl:404), (round(::Type{T}, x::Rational{Bool}, r::RoundingMode) where T>:Missing in Base at missing.jl:117, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesUp}) where T in Base at rational.jl:426), (round(::Type{T}, x::Rational, r::RoundingMode) where T>:Missing in Base at missing.jl:116, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:Nearest}) where T in Base at rational.jl:424), (round(::Type{T}, x::Rational, r::RoundingMode) where T>:Missing in Base at missing.jl:116, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesAway}) where T in Base at rational.jl:425), (round(::Type{T}, x, r::RoundingMode) where T>:Missing in Base at missing.jl:114, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:NearestTiesAway}) where {T, Tr} in Base at rational.jl:390), (round(::Type{T}, x::Rational{Bool}, r::RoundingMode) where T>:Missing in Base at missing.jl:117, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesAway}) where T in Base at rational.jl:425), (round(::Type{T}, x::Rational, r::RoundingMode) where T>:Missing in Base at missing.jl:116, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:Nearest}) where {T, Tr} in Base at rational.jl:374), (round(::Type{T}, x::Rational{Bool}, r::RoundingMode) where T>:Missing in Base at missing.jl:117, round(::Type{T}, x::Rational{Tr}, ::RoundingMode{:NearestTiesAway}) where {T, Tr} in Base at rational.jl:390), (round(::Type{T}, x::Rational, r::RoundingMode) where T>:Missing in Base at missing.jl:116, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesUp}) where T in Base at rational.jl:426), (round(::Type{T}, x, r::RoundingMode) where T>:Missing in Base at missing.jl:114, round(::Type{T}, x::Rational{Bool}, ::RoundingMode{:NearestTiesUp}) where T in Base at rational.jl:426)] == Any[]
```

amilsted committed:
Fix new round() ambiguities with Missing and Rational

This is kind of horrible, but I think it will work...

PS: Why isn't `round(Int, 3//4, RoundDown)` defined, while `floor(Int, 3//4)` is?

amilsted commented:
Unfortunately that didn't do it. Does anyone know how best to resolve this?

A though I had: Are the `Rational` versions of these functions perhaps too specific? Would these ambiguity problems go away if there was just a `round(::Type{T}, x::Rational{Tr}, r::RoundingMode) where {T,Tr}` that called some internal functions specialized to the particular `RoundingMode` types?

nalimilan commented:
Yes, having a common `round(::Type{T}, ::Rational, ::RoundingMode)` method which will then dispatch on an internal `_round` would probably help. Note that removing the type parameter of `Rational{Tr}` should also reduce ambiguities. If you do that you can probably remove some of the existing methods which are there just to avoid ambiguities with `Rational{Tr}`.

The other approach is to use `Rational{Tr}` instead of just `Rational` in missings.jl, but duplicating all of these methods isn't ideal.

amilsted committed:
Reduce Rational pollution of round() methods

The specification to `RoundingMode`s made it difficult to define `round()` for `Missing`.

amilsted committed:
Remove now-unneeded Missing,Rational methods

amilsted committed:
Fix syntax error in Missing test

amilsted committed:
Simplify Missing rounding tests

Remove duplicate tests but still test elementwise to help ensure we stay in sync with Float methods in the future. Add some comments to explain what is going on.

amilsted committed:
Remove redundant rounded_array

`is_equal(test_array, rounded_array) == true`, so just use `test_array`.

nalimilan commented on base/missing.jl:
```diff
@@ -107,16 +107,25 @@ max(::Missing, ::Any)     = missing
+round(::Type{T}, x::Any, r::RoundingMode=RoundNearest) where {T>:Missing} = round(nonmissingtype(T), x, r)
+# to fix ambiguities
+round(::Type{T}, x::Rational, r::RoundingMode=RoundNearest) where {T>:Missing} = round(nonmissingtype(T), x, r)
+round(::Type{T}, x::Rational{Bool}, r::RoundingMode=RoundNearest) where {T>:Missing} = round(nonmissingtype(T), x, r)
```
Is this one still needed?

Otherwise, looks good to me.

amilsted replied:
They're still needed, according to ambiguity test failure below.

nalimilan replied:
Note I said "this one", i.e. the `Rational{Bool}` method which has no equivalent in the loop below.

amilsted replied:
Yes. That one is still needed.

amilsted replied:
I had a typo in my tests which led me to miss the ambiguity. I see it now. Would be nice if it could be removed without having to mention Rational in missing.jl, but I don't see a way.

amilsted committed:
Remove Missing, Rational specializations

My (possibly incomplete) tests suggest these are no longer needed.

amilsted committed:
Restore Rational, Missing rounding functions

Still needed these to avoid ambiguities after all.

nalimilan reviewed (approved):
Thanks!

amilsted commented:
No worries. Was a lot more work than I originally expected! Will all these commits show up in the history if this gets accepted? It's a bit of a mess because I did it all on Github...

fredrikekre commented:
No, they will be squashed by the one that merges.

amilsted commented:
Yay.
